lib/core/src/Cardano/Wallet/Address/Pool.hs

```diff
@@ -0,0 +1,216 @@
+prop_generator Pool{generator,addresses} =
+    and $ Map.mapWithKey isGenerated addresses
+  where
+    isGenerated addr (ix,_) = generator ix == addr
+
+-- | Internal invariant: The pool satisfies all invariants above.
+prop_consistent :: (Ord ix, Enum ix, Eq addr) => Pool addr ix -> Bool
+prop_consistent p =
+    all ($ p) [prop_sequence, prop_gap, prop_fresh, prop_generator]
+
+{-------------------------------------------------------------------------------
+    Pretty printing
+-------------------------------------------------------------------------------}
+instance Buildable (Pool addr ix) where
+    build pool = "AddressPool "
+        <> "{ " <> build (size pool) <> " addresses"
+        <> ", gap = " <> build (gap pool)
+        <> "}"
+
+instance (Show addr, Show ix) => Show (Pool addr ix) where
+    show pool = "AddressPool"
+        <> "{ generator = <<function>>"
+        <> ", gap = " <> show (gap pool)
+        <> ", addresses = " <> show (addresses pool)
+        <> "}"
+
+{-------------------------------------------------------------------------------
+    Address Pool, operations
+-------------------------------------------------------------------------------}
+-- | Create a new address pool.
+new :: (Ord addr, Enum ix) => (ix -> addr) -> Int -> Pool addr ix
+new generator gap
+    = ensureFresh (toEnum 0) $ Pool{ generator, gap, addresses = Map.empty }
+
+-- | Replace the collection of addresses in a pool,
+-- but only if this collection satisfies the necessary invariants
+-- such as 'prop_sequence' etc.
+load
+    :: (Ord addr,  Ord ix, Enum ix)
+    => Pool addr ix -> Map addr (ix,AddressState) -> Maybe (Pool addr ix)
+load addrs pool0 = if prop_consistent pool then Just pool else Nothing
+  where pool = loadUnsafe addrs pool0
+
+-- | Replace the collection of addresses in a pool,
+-- but skips checking the invariants.
+loadUnsafe :: Pool addr ix -> Map addr (ix,AddressState) -> Pool addr ix
+loadUnsafe pool addrs = pool{ addresses = addrs }
+
+-- | Remove all previously discovered addresses,
+-- i.e. create a new pool with the same 'generator' and 'gap' as the old pool.
+clear :: (Ord addr, Enum ix) => Pool addr ix -> Pool addr ix
+clear Pool{generator,gap} = new generator gap
+
+-- | Look up an address in the pool.
+lookup :: Ord addr => addr -> Pool addr ix -> Maybe ix
+lookup addr Pool{addresses} = fst <$> Map.lookup addr addresses
+
+-- | Number of addresses cached in the pool.
+size :: Pool addr ix -> Int
+size = Map.size . addresses
+
+-- | Given an index, retrieve the next index that is still in the pool.
+next :: Enum ix => Pool addr ix -> ix -> Maybe ix
+next Pool{addresses} ix = let jx = succ ix in
+    if fromEnum jx >= Map.size addresses then Nothing else Just jx
+
+-- | Update an address to the 'Used' status
+-- and create new 'Unused' addresses in order to satisfy 'prop_fresh'.
+--
+-- Does nothing if the address was not in the pool.
+update :: (Ord addr, Enum ix) => addr -> Pool addr ix -> Pool addr ix
```

Comment:
if we update some address from Unused -> Used then we need to make sure "gap" invariant is not violated after this operation. We need to remember that we can have this two cases:
```
Used Unused Used Unused (gap times) -> change ix=3 ->  Used Unused Used Used Unused (gap times)
Used Unused Used Unused (gap times) -> change ix=1 -> Used Used Used Unused (gap times)
```
in the first case we are extending pool with one `Unused -> Used` flip, in the second we have just flip, no extension. This need to be tested!

Reply:
It took a few paper examples for me to prove to myself that `ensureFresh` works.

Reply:
> It took a few paper examples for me to prove to myself that ensureFresh works.

🤔 The key idea is that `[ix..to]` are the correct indices for which we need to generate addresses in order to satisfy `prop_fresh`. Using `fresh` instead of `ix` is an optimization. Here, we know that `fresh > ix`, and that indices before `fresh` are already in the map, so the range `[fresh..to]` is a subrange of `[ix..to]`.

But I think I failed to document a precondition on `ensureFresh`, namely that `jx` is the *last* 'Used' index. Will update.